src/main/haskell/kore/src/Kore/Unification/UnificationSolution.hs

```diff
@@ -0,0 +1,136 @@
+{-|
+Module      : Kore.Unification.UnificationSolution
+Description : Datastructures and functionality for performing unification on
+              Pure patterns
+Copyright   : (c) Runtime Verification, 2018
+License     : NCSA
+Maintainer  : vladimir.ciobanu@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+module Kore.Unification.UnificationSolution where
+
+import Data.Functor.Foldable
+import Kore.AST.Common
+import Kore.AST.PureML
+import Kore.AST.MLPatterns
+import Kore.IndexedModule.MetadataTools
+import Kore.Step.StepperAttributes
+       ( StepperAttributes )
+import Kore.Step.PatternAttributes
+       ( FunctionalProof (..) )
+
+type UnificationSubstitution level variable
+    = [(variable level, PureMLPattern level variable)]
+
+-- |'UnificationSolution' describes the solution of an unification problem,
+-- consisting of the unified term and the set of constraints (equalities)
+-- obtained during unification.
+data UnificationSolution level variable = UnificationSolution
+    { unificationSolutionTerm        :: !(PureMLPattern level variable)
+    , unificationSolutionConstraints :: !(UnificationSubstitution level variable)
+    }
+  deriving (Eq, Show)
+
+-- |'mapSubstitutionVariables' changes all the variables in the substitution
+-- with the given function.
+mapSubstitutionVariables
+    :: (variableFrom level -> variableTo level)
+    -> UnificationSubstitution level variableFrom
+    -> UnificationSubstitution level variableTo
+mapSubstitutionVariables variableMapper =
+    map (mapVariable variableMapper)
+  where
+    mapVariable
+        :: (variableFrom level -> variableTo level)
+        -> (variableFrom level, PureMLPattern level variableFrom)
+        -> (variableTo level, PureMLPattern level variableTo)
+    mapVariable
+        mapper
+        (variable, patt)
+      =
+        (mapper variable, mapPatternVariables mapper patt)
+
+-- |'unificationSolutionToPurePattern' packages an unification solution into
+-- a 'CommonPurePattern' by transforming the constraints into a conjunction of
+-- equalities and conjoining them with the unified term.
+unificationSolutionToPurePattern
+    :: SortedVariable variable
+    => MetadataTools level StepperAttributes
+    -> UnificationSolution level variable
+    -> PureMLPattern level variable
+unificationSolutionToPurePattern tools ucp =
+    case unificationSolutionConstraints ucp of
+        [] -> unifiedTerm
+        (constraint:constraints) ->
+            andPat unifiedTerm (foldr andEquals (equals constraint) constraints)
+  where
+    resultSort =
+        getPatternResultSort (sortTools tools) (project unifiedTerm)
+    unifiedTerm = unificationSolutionTerm ucp
+    andEquals = andPat . equals
+    andPat first second =
+        Fix $ AndPattern And
```

Comment:
Use [`Kore.ASTUtils.SmartPatterns.And_`](https://github.com/kframework/kore/blob/master/src/main/haskell/kore/src/Kore/ASTUtils/SmartPatterns.hs#L51).

Reply:
Ended up removing the function altogether.